x-pack/plugins/enterprise_search/public/applications/app_search/components/relevance_tuning/relevance_tuning_logic.ts

```diff
@@ -155,4 +287,270 @@ export const RelevanceTuningLogic = kea<
+      const { search_fields: searchFields } = searchSettings;
+
+      actions.setSearchSettings({
+        ...searchSettings,
+        boosts: searchSettings.boosts,
+        search_fields: {
+          ...omit(searchFields, name),
+          ...(!disableField && { [name]: { weight: 1 } }),
+        },
+      });
+      actions.getSearchResults();
+    },
+    updateFieldWeight: ({ name, weight }) => {
+      const { searchSettings } = values;
+      const { search_fields: searchFields } = searchSettings;
+
+      actions.setSearchSettings({
+        ...searchSettings,
+        boosts: searchSettings.boosts,
+        search_fields: {
+          ...searchFields,
+          [name]: {
+            ...searchFields[name],
+            weight: Math.round(weight * 10) / 10,
+          },
+        },
+      });
+      actions.getSearchResults();
+    },
+    addBoost: ({ name, type }) => {
+      const { searchSettings } = values;
+      const { boosts } = searchSettings;
+      const emptyBoost = { type, factor: 1, newBoost: true };
+      let boostArray;
+
+      if (Array.isArray(boosts[name])) {
+        boostArray = boosts[name].slice();
+        boostArray.push(emptyBoost);
+      } else {
+        boostArray = [emptyBoost];
+      }
+
+      actions.setSearchSettings({
+        ...searchSettings,
+        boosts: {
+          ...boosts,
+          [name]: boostArray,
+        },
+      });
+    },
+    deleteBoost: ({ name, index }) => {
+      if (window.confirm(DELETE_CONFIRMATION_MESSAGE)) {
+        const { searchSettings } = values;
+        const { boosts } = searchSettings;
+        const boostsRemoved = boosts[name].slice();
+        boostsRemoved.splice(index, 1);
+        const updatedBoosts = { ...boosts };
+
+        if (boostsRemoved.length > 0) {
+          updatedBoosts[name] = boostsRemoved;
+        } else {
+          delete updatedBoosts[name];
+        }
+
+        actions.setSearchSettings({
+          ...searchSettings,
+          boosts: updatedBoosts,
+        });
+        actions.getSearchResults();
+      }
+    },
+    updateBoostFactor: ({ name, index, factor }) => {
+      const { searchSettings } = values;
+      const { boosts } = searchSettings;
+      const updatedBoosts = cloneDeep(boosts[name]);
+      updatedBoosts[index].factor = Math.round(factor * 10) / 10;
+
+      actions.setSearchSettings({
+        ...searchSettings,
+        boosts: {
+          ...boosts,
+          [name]: updatedBoosts,
+        },
+      });
+      actions.getSearchResults();
+    },
+    updateBoostValue: ({ name, boostIndex, valueIndex, value }) => {
+      const { searchSettings } = values;
+      const { boosts } = searchSettings;
+      const updatedBoosts: Boost[] = cloneDeep(boosts[name]);
+      const existingValue = updatedBoosts[boostIndex].value;
+      if (existingValue === undefined) {
+        updatedBoosts[boostIndex].value = [value];
+      } else {
+        existingValue[valueIndex] = value;
+      }
```

Comment:
[**Not** a change request, just me thinking out loud]

There's a lot going on in w/ these boost listeners that I think generally makes this file a little hard for me to initially parse. I know you said you weren't really looking to make significant refactors to current code, but I think in this scenario it would have made unit tests easier to write/read (avoiding 1000+ line of tests for a single logic file) and might be worth considering.

 If I had to make a suggestion on how to make it more approachable, it would be 1 of 2 ways:

1. Take out all boost-related logic (slicing arrays, deep cloning, etc.) to a separate sub-folder / utility file. All the listener then does is essentially this:

```js
updateBoostFactor: ({ name, index, factor }) => {
  const { searchSettings: { boosts } } = values;
  const updatedBoosts = someUtility({ boosts, name, index, factor });

  actions.setSearchSettings({ ...searchSettings, boosts: updatedBoosts });
},
updateBoostValue: ({ name, boostIndex, valueIndex, value }) => {
  const { searchSettings: { boosts } } = values;
  const updatedBoosts = anotherUtility({ boosts, name, boostIndex, valueIndex, value });

  actions.setSearchSettings({ ...searchSettings, boosts: updatedBoosts });
},
```

2. Totally separate Logic file for boosts that has its own boost state, listeners, etc. (might be more effort than it's worth, I think #1 appeals to me slightly more)

Again, not a change request, but since we do _have_ to write unit tests, it's worth considering that our old code is not optimized for unit testing, & while we don't need to refactor everything, refactors that break up code into more modular pieces is both easier to review **and** write unit tests for.

Reply:
This is great feedback Constance, thanks. This is definitely overly verbose and repetitive, and your suggestion is great. I would love to clean this up but I think I'm going to move on from this for now. As the Relevance Tuning work progresses and I get back into the logic file, maybe I'll be able to find a moment or two to clean this up.

Reply:
Yeah totally fair, thanks a ton Jason!